compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,494 @@
+        usePar = se->parentExpr;
+        if(CallExpr* ce = toCallExpr(usePar)) {
+          if( !(ce->isPrimitive(PRIM_ADDR_OF) ||
+                ce->isPrimitive(PRIM_ARRAY_GET) ||
+                ce->isPrimitive(PRIM_GET_MEMBER) ||
+                ce->isPrimitive(PRIM_DEREF) ||
+                ce->isPrimitive(PRIM_GET_MEMBER_VALUE) ||
+                (ce->isPrimitive(PRIM_MOVE) && 
+                 ce->get(1)->typeInfo() !=
+                 ce->get(2)->typeInfo()))) {
+            use = se;
+          }
+        }
+      }
+      else {
+        return false;
+      }
+      if(use) {
+        *useOut = use;
+      }
+      else {
+        return false;
+      }
+
+      //here I have my valid use and def
+      //so far we checked specific cases for them individually, now check if
+      //there is anything wrong with them as a pair
+
+      //this issue feels too specific, maybe there is
+      //safer/better/more general way of doing this check
+      //
+      //for reference test that caused this was:
+      //test/modules/standard/FileSystem/bharshbarg/filer
+      //
+      //The issue seemed to be yielding string from an iterator
+      if(CallExpr* useParentCe = toCallExpr(usePar)) {
+        if(useParentCe->isPrimitive(PRIM_FTABLE_CALL)) {
+          if(argMustUseCPtr(def->typeInfo())){
+            return false;
+          }
+        }
+      }
+
+      // we have to protect repeatedly evaluated statements of loops from
+      // expensive and/or unsafe CallExprs
+      if(CallExpr* defCe = toCallExpr(def)){
+        if(! isNonEssentialPrimitive(defCe)) { //nonessential primitives are safe
+          if(LoopStmt* enclLoop = LoopStmt::findEnclosingLoop(use)) {
+            if(CForLoop* enclCForLoop = toCForLoop(enclLoop)) {
+              if(enclCForLoop->testBlockGet()->contains(ce) ||
+                  enclCForLoop->incrBlockGet()->contains(ce)) {
+                return false;
+              }
+            }
+            else if(enclLoop->isWhileStmt() || 
+                enclLoop->isDoWhileStmt() || 
+                enclLoop->isWhileDoStmt()) {
+              if(toWhileStmt(enclLoop)->condExprGet()->contains(ce)) {
+                return false;
+              }
+            }
+          }
+        }
+      }
+      return true;
+    }
+  }
+  return false;
+}
+
+bool isExprSafeForReorder(Expr * e) {
+  if(safeExprCache.count(e) > 0) {
+    return safeExprCache[e];
+  }
+  if(CallExpr* ce = toCallExpr(e)) {
+    if(!ce->isPrimitive()) {
+      FnSymbol* fnSym = ce->theFnSymbol();
+
+      const bool cachedGlobalManip = globalManipFuncCache.count(fnSym);
+
+      if(cachedGlobalManip) {
+        if(globalManipFuncCache[fnSym]) {
+          return false;
+        }
+      }
+      for_formals(formal, fnSym) {
+        if(isReferenceType(formal->typeInfo())) {
+          safeExprCache[e] = false;
+          return false;
+        }
+      }
+
+      std::vector<BaseAST*> asts;
+      collect_asts(fnSym->body, asts);
```

Comment:
I suggest replacing the above collect_asts with collectSymExprs and moving it inside the if(!cachedGlobalManip) below, to save some compilation time.